src/Build/Utilities/EngineFileUtilities.cs

```diff
@@ -179,6 +338,152 @@ FileMatcher fileMatcher
             return fileList;
         }
 
+        private static void LogDriveEnumerationWarningWithTargetLoggingContext(TargetLoggingContext targetLoggingContext, IElementLocation includeLocation, bool excludeFileSpecIsUnusable, bool disableExcludeDriveEnumerationWarning, string fileSpec)
+        {
+            // Both condition lines are necessary to skip for the first GetFileListEscaped call
+            // and reach for the GetFileListUnescaped call when the wildcarded Exclude attribute results
+            // in a drive enumeration. Since we only want to check for the Exclude
+            // attribute here, we want to ensure that includeLocation is null - otherwise,
+            // Include wildcard attributes for the GetFileListEscaped calls would falsely appear
+            // with the Exclude attribute in the logged warning.
+            if (((!excludeFileSpecIsUnusable) && (!disableExcludeDriveEnumerationWarning)) ||
```

Comment:
nit:
The extra parentheses here are actually making it harder for me to figure out what's grouped. I think you're saying that disableExcludeDriveEnumerationWarning could be true, and you'd still log a warning here if includeLocation is null, correct? That feels wrong. Also, shouldn't you propagate that condition to other conditions?

Reply:
Right, I would want to log the warning if includeLocation is null, since I do not want Include wildcard attributes for the GetFileListEscaped calls to falsely appear with the Exclude attribute in the message that appears with the logged warning. It seemed more foolproof to me based on the ordering of calls in ItemGroupIntrinsicTask along with the combination of how exclude elements are detected in FileMatcher. The reason why I don't propagate that condition (I assume you are referring to `includeLocation == null`) to other conditions is because I want exclude (or project) elements to be captured as well.

Reply:
It's not as necessary to check the ElementLocation(s) for several of those, because the logic in FileMatcher has already detected a drive enumerating wildcard at this point.

Reply:
To specify more of what I meant by that last statement, I think the logic is rigid to the point where it will know what codepath the invalid filespec came from to properly log the right warning or right exception, but if changes were made to each of those codepaths, those changes may need to be reflected in `EngineFileUtilities` as well.